### PR TITLE
feat: Enable new

### PR DESCRIPTION
**Pull request by jason-famedly**
Part of #26 

Tasks:
* [ ] Parse `tim-type` as a new setting option. Update README
* [ ] Default `tim-type` to 'Pro' mode
* [ ] 'ePA' mode blocks all local invites
* [ ] Start saving comprehensive results from the parsed Federation List
* [ ] Implement serving new endpoint `/isInsurance` based on Federation Data
* [ ] Implement serving new endpoint `/findByIk` based on Federation Data
* Test new configuration setting:
   * [ ] Parsing is case-insensitive, defaults to 'Pro' mode, and(if incorrectly set) raises an Error
* 'ePA' mode 
   * [ ] Test all local invites are blocked
   * [ ] Test outgoing remote invites are blocked if remote server is an 'Insurance' type(What do we do for incoming remote invites from same? I guess they are allowed so Insurance can contact a patient)
   * [ ] Test new `/isInsurance` and `/findByIk` return expected error code 401
* 'Pro' mode
   * [ ] Test all local invites are not blocked
   * [ ] Test outgoing remote invites are allowed
   * [ ] Test new `/isInsurance` and `/findByIk` return correct data

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -37,6 +37,7 @@ modules:
         federation_localization_url: "https://localhost:8080/localization", # Full url where to fetch the federation localization from, required. Should be the same host as federation list.
         federation_list_client_cert: "tests/certs/client.pem", # path to a pem encoded client certificate for mtls, required if federation list url is https
         gematik_ca_baseurl: "https://download-ref.tsl.ti-dienste.de/", # the baseurl to the ca to use for the federation list, required
+        tim-type: "epa" or "pro", # Patient or Professional mode, optional. Defaults to "pro"
```
Comment (nico-famedly) [inline README.md:40]:
```suggestion
        tim-type: "epa" or "pro", # Patient/Insurance or Professional mode, optional. Defaults to "pro"
```

I think the epa mode is really to run the insurance services to give patients access to TIM, so we probably should name them as well?

Comment (jason-famedly) [reply README.md:40]: We can, but only patients can be insured. I felt it was implied terminology. If it's a doctor being seen, he's a 'patient' in that context and has to go thru his insurance to connect! (Which, to be fair, I see as not happening when they can take the 'already established route and just message another doctor)

File: synapse_invite_checker/invite_checker.py [hunk 2]
```diff
@@ -296,13 +321,17 @@ def _load_cert_b64(self, cert: str) -> X509:
         return load_certificate(FILETYPE_ASN1, base64.b64decode(cert))
 
     @cached(cache=TTLCache(maxsize=1, ttl=60 * 60))
-    async def fetch_federation_allow_list(self) -> tuple[set[str], FederationList]:
+    async def fetch_federation_allow_list(
+        self,
+    ) -> tuple[set[str], set[str], FederationList]:
```
Comment (nico-famedly) [inline synapse_invite_checker/invite_checker.py:326]: This should really be a class at this point, this is getting hard to parse :)

File: synapse_invite_checker/invite_checker.py [hunk 3]
```diff
@@ -179,14 +183,17 @@ def __init__(self, config: InviteCheckerConfig, api: ModuleApi):
 
         # The TiMessengerInformation API resource
         MessengerInfoResource(self.config).register(self.resource)
+        MessengerIsInsuranceResource(
+            self.config, self.is_domain_insurance_with_retry
```
Comment (nico-famedly) [inline synapse_invite_checker/invite_checker.py:187]: Imo it is easier to read, if you call that directly on the InviteChecker instead of passing in a callback?

File: README.md (hunk 1 above)
Comment (nico-famedly) [reply README.md:40]: This should also mention, that this will be required in the future, so people really should set it?

File: synapse_invite_checker/rest/messenger_info.py [hunk 4]
```diff
@@ -42,11 +43,36 @@ def __init__(self, config: InviteCheckerConfig):
...
+        self.PATTERNS = tim_info_patterns("/v1/server/isInsurance$")
+
+    async def on_GET(self, request: SynapseRequest) -> tuple[int, JsonDict]:
+        if self.config.tim_type == TimType.EPA:
+            return 401, {
+                "errorCode": "thing",
+                "errorMessage": "Unauthorized to access for ePA backend",
+            }
+
+        server_name = parse_string(request, "serverName")
```
[the remark this answers is not in the document]
Comment (jason-famedly) [reply synapse_invite_checker/rest/messenger_info.py:74]: Yes, good catch

File: tests/test_local_invites.py [hunk 5]
```diff
@@ -35,7 +42,7 @@ def test_invite_to_dm(self) -> None:
         room_id = self.helper.create_room_as(
             self.user_a, is_public=False, tok=self.access_token
         )
-        assert room_id, "Room created"
+        assert room_id, "Room not created"
```
Comment (nico-famedly) [inline tests/test_local_invites.py:45]: Wouldn't now the error be: "Assertion Room not created failed"? That seems backwards to me.

File: synapse_invite_checker/invite_checker.py (hunk 3 above)
Comment (jason-famedly) [reply synapse_invite_checker/invite_checker.py:187]: Maybe, but this spells out exactly what function is associated with that rest endpoint without looking at the resource construction. I didn't want to just pass the whole InviteChecker in when all that is needed from it is the config and the one function to call

File: synapse_invite_checker/invite_checker.py (hunk 2 above)
Comment (jason-famedly) [reply synapse_invite_checker/invite_checker.py:326]: I'll consider it in a following round. There are a couple different angles and variables to take into account and it needs discussion

File: synapse_invite_checker/rest/messenger_info.py [hunk 6]
```diff
@@ -42,11 +43,36 @@ def __init__(self, config: InviteCheckerConfig):
...
+        self.is_insurance_cb = is_insurance_cb
+
+        self.PATTERNS = tim_info_patterns("/v1/server/isInsurance$")
+
+    async def on_GET(self, request: SynapseRequest) -> tuple[int, JsonDict]:
+        if self.config.tim_type == TimType.EPA:
+            return 401, {
+                "errorCode": "thing",
+                "errorMessage": "Unauthorized to access for ePA backend",
+            }
```
[the remark this answers is not in the document]
Comment (jason-famedly) [reply synapse_invite_checker/rest/messenger_info.py:72]: I think I got held up there by going to look for the verbiage on the `errorMessage`, not finding it and instead discovering several other rabbit holes of semi-related things. Good catch

File: tests/test_local_invites.py [hunk 7]
```diff
@@ -121,3 +128,122 @@ def test_invite_to_group_without_dm_event(self) -> None:
...
+    These tests operate under the assumption that invites are processed after room
+    creation has occurred. Invites during room creation will be tested separately
+    """
+
+    def prepare(self, reactor: MemoryReactor, clock: Clock, homeserver: HomeServer):
+        super().prepare(reactor, clock, homeserver)
+        self.user_a = self.register_user("a", "password")
+        self.access_token = self.login("a", "password")
+        self.user_b = self.register_user("b", "password")
+        self.user_c = self.register_user("c", "password")
```
Comment (jason-famedly) [inline tests/test_local_invites.py:144]: That reminds me tho, need to fix these. These users can't exist on an insurance domain(I think, I might be mis-remembering)

File: synapse_invite_checker/invite_checker.py [hunk 8]
```diff
@@ -362,21 +422,27 @@ async def user_may_invite(
...
+        if (
+            self.config.tim_type == TimType.EPA
+            and await self.is_domain_insurance_with_retry(inviter_domain)
+        ):
+            # Per AF_10233: Deny incoming remote invites if in ePA mode(which means the
+            # local user is an 'insured') and if the remote domain is type 'insurance'.
+            logger.warning(
+                "Discarding invite from remote insurance domain: %s", inviter_domain
+            )
+            return errors.Codes.FORBIDDEN
```
[the remark this answers is not in the document]
Comment (jason-famedly) [reply synapse_invite_checker/invite_checker.py:445]: You know what? You may be right. I was passing that point of view on the parent issue, but the spec *does* say to check both us and them.
https://gemspec.gematik.de/docs/gemSpec/gemSpec_TI-M_ePA/latest/#A_25705

I'll get this sorted out soon

**Comment by jason-famedly** [comment]
squashing up to this point, so new work will be after that

File: synapse_invite_checker/invite_checker.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (nico-famedly) [reply synapse_invite_checker/invite_checker.py:467]: In the end the other side will check the federation list when validating the invite. As such if the admin sets something else, it will break randomly and hard to diagnose. That is why I would make a noisy warning on startup if the setting is different than what we would expect from the federation list.

Now if we should check epa or the list when doing the actual checks, I don't know. I would probably prioritize the list, but as long as the admin sets the setting correctly, it shouldn't matter? Which is why I think warning on startup makes sense?

File: synapse_invite_checker/invite_checker.py [hunk 9]
```diff
@@ -337,14 +366,45 @@ async def fetch_federation_allow_list(self) -> tuple[set[str], FederationList]:
...
+    async def is_domain_insurance_with_retry(self, domain: str) -> bool:
+        _, insurance_domains, _ = await self.fetch_federation_allow_list()
+        if domain in insurance_domains:
+            return True
+
+        # The domain wasn't found but the list may have changed since the last look.
+        # Re-fetch the list and try again
+        self.fetch_federation_allow_list.cache_clear()
+        _, insurance_domains, _ = await self.fetch_federation_allow_list()
+        return domain in insurance_domains
```
[the remark this answers is not in the document]
Comment (jason-famedly) [reply synapse_invite_checker/invite_checker.py]: See what you think about [413381a](https://github.com/famedly/synapse-invite-checker/pull/37/commits/413381a2ee750c7deeae497d0ffc2d47c2704be3)

File: synapse_invite_checker/invite_checker.py (hunk 2 above)
Comment (jason-famedly) [reply synapse_invite_checker/invite_checker.py]: I think I did that in [413381a](https://github.com/famedly/synapse-invite-checker/pull/37/commits/413381a2ee750c7deeae497d0ffc2d47c2704be3) but it may not have been what you meant

File: synapse_invite_checker/invite_checker.py [hunk 10]
```diff
@@ -296,13 +343,17 @@ def _load_cert_b64(self, cert: str) -> X509:
...
         """
         Fetch the raw data for the federation list, verify it is authentic and parse
         the data into a usable format
 
-        Returns: a tuple of a set of the domains from the list and the complete
-            validated list(for testing)
+        Returns:
+            a 3-tuple of a set of the domains from the list,
+            a set of domains that are insurance,
+            and the complete validated list(for testing)
```
Comment (nico-famedly) [inline synapse_invite_checker/invite_checker.py:356]: This seems wrong? :)

File: synapse_invite_checker/types.py [hunk 11]
```diff
@@ -61,3 +62,28 @@ class FederationList(BaseModel):
...
+    @computed_field
+    @cached_property
+    def _insurance_domains_on_list(self) -> list[str]:
+        return [
+            domain_data.domain
+            for domain_data in self.domainList
+            if domain_data.isInsurance
+        ]
+
+    def allowed(self, domain: str) -> bool:
```
Comment (nico-famedly) [inline synapse_invite_checker/types.py:80]: I know I used that term out of laziness, but probably should be "federation_allowed" or have some documentation?

File: tests/test_local_invites.py [hunk 12]
```diff
@@ -12,21 +12,32 @@
...
+    """
+
     def prepare(self, reactor: MemoryReactor, clock: Clock, homeserver: HomeServer):
         super().prepare(reactor, clock, homeserver)
         self.user_a = self.register_user("a", "password")
         self.access_token = self.login("a", "password")
         self.user_b = self.register_user("b", "password")
         self.user_c = self.register_user("c", "password")
 
+        # @d:test is none of these things and should be just a 'User'
```
Comment (nico-famedly) [inline tests/test_local_invites.py:37]: None of what? Not a practitioner?

File: synapse_invite_checker/invite_checker.py (hunk 10 above)
Comment (jason-famedly) [reply synapse_invite_checker/invite_checker.py:356]: Good catch, fixed

File: synapse_invite_checker/types.py [hunk 13]
```diff
@@ -61,3 +62,28 @@ class FederationList(BaseModel):
     )
 
     domainList: list[FederationDomain]  # noqa: N815
+
+    @computed_field
+    @cached_property
+    def _domains_on_list(self) -> list[str]:
```
[the remark this answers is not in the document]
Comment (jason-famedly) [reply synapse_invite_checker/types.py:68]: I think they can be, the sequence itself doesn't have to have a concrete order. Fixed

File: synapse_invite_checker/types.py (hunk 11 above)
Comment (jason-famedly) [reply synapse_invite_checker/types.py:80]: I added some docstrings

File: tests/test_remote_invites.py [hunk 14]
```diff
@@ -162,24 +236,350 @@ async def test_invite_from_remote_to_local_org(self) -> None:
...
+        if ret == NOT_SPAM:
+            return NOT_SPAM
+        return ret[0]  # return first code instead of all of them to make assert easier
+
+    async def test_invite_from_remote_not_on_fed_list(self) -> None:
+        """Tests that an invite from a remote server not in the federation list gets denied"""
+
+        channel = self.make_request(
+            "PUT",
+            "/_synapse/client/com.famedly/tim/v1/contacts",
```
Comment (nico-famedly) [inline tests/test_remote_invites.py:344]: I'm not sure the contacts API should be available in EPA mode?

File: tests/test_local_invites.py (hunk 12 above)
Comment (jason-famedly) [reply tests/test_local_invites.py:37]: Yes. So like a nurse or a doctor's secretary. From what I can understand from the spec these people are considered Users. Like an office manager or something. I updated the comment to add what each user a-c is since that was a fair bit of missing context(In the other similar places too)

File: tests/test_remote_invites.py [hunk 15]
```diff
@@ -162,24 +236,350 @@ async def test_invite_from_remote_to_local_org(self) -> None:
...
+        self.helper.auth_user_id = self.user_d
+
+    def default_config(self) -> dict[str, Any]:
+        conf = super().default_config()
+        assert "modules" in conf, "modules missing from config dict during construction"
+
+        # There should only be a single item in the 'modules' list, since this tests that module
+        assert len(conf["modules"]) == 1, "more than one module found in config"
+
+        conf["modules"][0].setdefault("config", {}).update({"tim-type": "epa"})
```
Comment (nico-famedly) [inline tests/test_remote_invites.py:324]: does that work properly if our hs name is "test"?

File: tests/test_local_invites.py [hunk 16]
```diff
@@ -121,3 +146,136 @@ def test_invite_to_group_without_dm_event(self) -> None:
...
+            targ=self.user_d,
+            tok=self.access_token,
+            expect_code=200,
+        )
+
+
+class LocalEpaModeInviteTest(ModuleApiTestCase):
+    """
+    These tests operate under the assumption that invites are processed after room
+    creation has occurred. Invites during room creation will be tested separately
```
[the remark this answers is not in the document]
Comment (jason-famedly) [reply tests/test_local_invites.py:161]: I don't actually believe they do, at least when it comes the follow up it didn't seem to. I spelled this out in the context of the issue that mentioned changing the SDK to only do invites after room creation(even tho that was for invites of additional users after the fact)

File: tests/test_remote_invites.py (hunk 14 above)
Comment (jason-famedly) [reply tests/test_remote_invites.py:344]: It's a gray area for me. I thought that if I was a patient and I needed to contact my doctor who was not a publicly listed practitioner, then I needed their contact details?

File: tests/test_remote_invites.py (hunk 15 above)
Comment (jason-famedly) [reply tests/test_remote_invites.py:324]: It does because we just warn now if there is a conflict in our two sources of "truth"

File: tests/test_local_invites.py (hunk 12 above)
Comment (nico-famedly) [reply tests/test_local_invites.py:40]: You use such users in other tests, but not in these. So I think you just copy and pasted that user and don't need them here?

File: tests/test_local_invites.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (nico-famedly) [reply tests/test_local_invites.py:164]: I am pretty sure I tested this and the spam check endpoint does get invoked during room creation. The follow up is about something else, where multiple invites during room creation can't reliably return errors.

File: tests/test_remote_invites.py (hunk 14 above)
Comment (nico-famedly) [reply tests/test_remote_invites.py:347]: The contact API is about making yourself contactable. Patients don't need that and they can't use that to contact doctors either, but double check the spec I guess?

File: tests/test_remote_invites.py (hunk 15 above)
Comment (nico-famedly) [reply tests/test_remote_invites.py:327]: That's not really intended functionality though. I guess it works, but should have a comment then.

File: tests/test_local_invites.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jason-famedly) [reply tests/test_local_invites.py:164]: And yet doctors are allowed to invite additional users to a room after the fact. So, I guess I don't understand what you are asking?

File: tests/test_local_invites.py (hunk 12 above)
Comment (jason-famedly) [reply tests/test_local_invites.py:40]: Oh I have their access token. I can lose that, or I can write an additional iteration of tests to check that scenario. Which would you prefer?

File: tests/test_local_invites.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jason-famedly) [reply tests/test_local_invites.py:164]: > I am pretty sure I tested this and the spam check endpoint does get invoked during room creation. The follow up is about something else, where multiple invites during room creation can't reliably return errors.

You may have tested this, but it is not currently tested here

File: tests/test_remote_invites.py (hunk 14 above)
Comment (jason-famedly) [reply tests/test_remote_invites.py:347]: I spent(far to much) time examine the spec and could find nothing that spells out that a patient is not allowed to have a contact permission feature. I think the contact permissions here are incomplete though, I'm seeing a "Block All" but allow exceptions and the inverse is missing. Let's kick that further down the road as it is creeping into a larger problem.

File: synapse_invite_checker/invite_checker.py [hunk 17]
```diff
@@ -362,21 +434,26 @@ async def user_may_invite(
...
+        ):
+            # Per AF_10233: Deny incoming remote invites if in ePA mode(which means the
+            # local user is an 'insured') and if the remote domain is type 'insurance'.
+            logger.warning(
+                "Discarding invite from remote insurance domain: %s", inviter_domain
+            )
+            return errors.Codes.FORBIDDEN
 
         # Step 2, check invite settings
         contact = await self.store.get_contact(UserID.from_string(invitee), inviter)
```
[the remark this answers is not in the document]
Comment (jason-famedly) [reply synapse_invite_checker/invite_checker.py:457]: combine with other discussion around this area
https://github.com/famedly/synapse-invite-checker/pull/37#discussion_r1908582012

File: tests/test_remote_invites.py (hunk 15 above)
Comment (jason-famedly) [reply tests/test_remote_invites.py:333]: I added such a comment in both places I thought it was relevant

File: tests/test_local_invites.py (hunk 12 above)
Comment (jason-famedly) [reply tests/test_local_invites.py:39]: I just removed the access token for that user. It wasn't actually being used. May add tests for that later

File: synapse_invite_checker/invite_checker.py [hunk 18]
```diff
@@ -179,14 +185,22 @@ def __init__(self, config: InviteCheckerConfig, api: ModuleApi):
...
+        MessengerIsInsuranceResource(self.config, self.is_domain_insurance).register(
+            self.resource
+        )
 
         # Register everything at the root of our namespace/app, to avoid complicating
         # Synapse's regex http registration
         self.api.register_web_resource(BASE_API_PREFIX, self.resource)
+
+        # Install a callable that will be registered to run after the reactor has been
+        # completely setup and has started running.
```
Comment (nico-famedly) [inline synapse_invite_checker/invite_checker.py:197]: That part is obvious, if you want to document something, document why we are doing that, i.e.:

> We can only validate the EPA mode once the HS is running.

File: synapse_invite_checker/invite_checker.py [hunk 19]
```diff
@@ -223,8 +237,41 @@ def parse_config(config):
...
+            else:
+                msg = "`tim-type` setting is not a recognized value. Please fix."
+                raise ConfigError(msg)
+
         return _config
 
+    def after_startup(self) -> None:
+        """
+        Attach to the reactor, and after the reactor is fully set up and running
+        check the federation list to see if if the current server isInsurance
```
Comment (nico-famedly) [inline synapse_invite_checker/invite_checker.py:263]:
```suggestion
        To be called when the rector is running. Validates that the epa setting matches the insurance setting in the federation list.
```

File: synapse_invite_checker/invite_checker.py [hunk 20]
```diff
@@ -362,23 +434,33 @@ async def user_may_invite(
...
-                logger.warning("Discarding invite from domain: %s", inviter_domain)
-                return errors.Codes.FORBIDDEN
+        # Step 1a, check federation allow list
+        if not await self.is_domain_allowed(inviter_domain):
+            logger.warning("Discarding invite from domain: %s", inviter_domain)
+            return errors.Codes.FORBIDDEN
+
+        # Step 1b, check that both domains are not of type 'epa'
+        # TODO: stop trusting our local setting and instead check the federation list
+        #  data to see if *this* server is of type `isInsurance`
```
Comment (nico-famedly) [inline synapse_invite_checker/invite_checker.py:450]: Is this a todo? I thought we intentionally use the epa setting?

```suggestion
        # We warn during startup if the 'epa' setting and isInsurance are not equivalent, so we can use the setting to know, if we are an insurance.
```

Comment (nico-famedly) [reply synapse_invite_checker/invite_checker.py:450]: this seems to be repeated later, so you can probably merge those comments.

**Review by nico-famedly — DISMISSED** [review]
Code seems fine, some of the comments still are a bit redundant to me, but whatever :)

File: synapse_invite_checker/invite_checker.py (hunk 18 above)
Comment (jason-famedly) [reply synapse_invite_checker/invite_checker.py:197]: I'll remove it

File: synapse_invite_checker/invite_checker.py (hunk 19 above)
Comment (jason-famedly) [reply synapse_invite_checker/invite_checker.py:263]: Sounds good, I'll use that
